File: src/main/java/dev/mazurkiewicz/quizer/question/domain/port/QuestionService.java
package dev.mazurkiewicz.quizer.question.domain.port;

import dev.mazurkiewicz.quizer.question.domain.model.*;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class QuestionService {

    private final QuestionRepository repository;

    public Question getQuestion(QuestionId questionId) {
        return repository.getQuestionById(questionId);
    }

    public Question getRandomQuestion() {
        return repository.getRandomQuestion();
    }

    public QuestionInfo getQuestionInfo() {
        return repository.getQuestionInfo();

    }

    public AnswerResult checkAnswer(QuestionId questionId, SelectedAnswer selectedAnswer) {
        Question question = repository.getQuestionById(questionId);
        return question.checkAnswer(selectedAnswer);
    }
}

File: src/main/kotlin/io/mazurkiewicz/quizer/question/domain/port/QuestionService.kt
package io.mazurkiewicz.quizer.question.domain.port

import io.mazurkiewicz.quizer.question.domain.model.Question
import io.mazurkiewicz.quizer.question.domain.model.QuestionId
import io.mazurkiewicz.quizer.quiz.domain.port.QuizTemplateService

class QuestionService(
    private val questionRepository: QuestionRepository,
    private val templateService: QuizTemplateService
) {

    fun getQuestion(questionId: QuestionId): Question {
        return questionRepository.findQuestionById(questionId)
    }

    fun saveQuestion(question: Question) {
        question.validateAnswers()
        templateService.validateAccessToTemplate(question.template, question.author)
        questionRepository.saveQuestion(question)
    }
}
